### PR TITLE
Merge ardana scripts/playbooks from CI and QE

### PR DESCRIPTION
**Pull request by flaviodsr**
This change enable the deploy and testing of ardana, for both virtual and physical deployments, using the same code/steps whenever possible independent of the type of the environment.

Summary of ansible/scripts changes:
- public ssh keys are now managed by its respective role and ssh keys are stored in a dictionary at `group_vars/all/ssh_pub_keys.yml`
- removed Nuremberg hosts from the inventory as QE will not use those hosts due to hardware requirements
- restructured group_vars variables for common use on virtual and physical deployments
- simplified the inventory by removing unnecessary groups
- renamed QA hosts from `ardana-qeXXX` to `qeXXX`
- the bootstrap_deployer role was split in two roles (`setup_zypper_repos` and `bootstrap_ardana`)
- added role and playbook for cloning input model from the ardana-input-model repository
- added role and playbook for creating/deleting heat stacks
- added playbooks for bootstrapping physical and virtual nodes
- added playbook for collecting parameters from a jenkins job and converting them into ansible variables

Summary of JJB changes:
- renamed openstack-ardana-bootstrap to openstack-ardana-bootstrap-clm
- reorganized and cleaned up job parameters to correspond to the new parameters accepted by the merged ansible playbooks
- created a separate job for initializing the deployer for a physical cloud

Summary of Jenkinsfile changes:
- use the `load-job-params.yml` to collect all the parameters from the         Jenkins job and convert them into ansible variables. This removes the         need to use Jenkins environment variables to pass information between         stages and simplifies adding/removing parameters to/from the job.
- enabled proper coloring for the ansible logs
- archive artifacts from every job. For tempest, in particular, using         JUnit test results enables Jenkins to process them in detail, highlight         failing test cases and compare them with previous results, reporting         differences between subsequent runs.
- all of the pipeline stages are now calling ansible playbooks (with a couple of minor exceptions)

### CODE REVIEW

**Comment by JanZerebecki** [comment]
@stefannica did you intent to also remove the 'do not merge yet' label?

File: scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: ecp_os_stack
```
Comment (JanZerebecki) [inline scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py]: How is this different from the module it was forked from?
Why is there a need to have a copy here?

Comment (flaviodsr) [reply scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py]: The only difference is, due to the ECP API not being reliable (sometimes when creating a stack it returns HTTP 500), I have added a retry when the request returns 500.

**Comment by flaviodsr** [comment]
> This PR does way too many things in one go. That makes reviewing and addressing feedback much harder. See wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes for a detailed explanation.

I totally agree, and I am sorry about it. With the help of @stefannica we managed to break it at least in different commits, but I am afraid we cannot do more than this.

> @stefannica did you intent to also remove the 'do not merge yet' label?

I will remove the label as soon as I finish testing it.

**Comment by flaviodsr** [comment]
Tested on:

- gating-cloud9: https://ci.suse.de/blue/organizations/jenkins/openstack-ardana-flavio/detail/openstack-ardana-flavio/123/pipeline
- gating-cloud8: https://ci.suse.de/blue/organizations/jenkins/openstack-ardana-flavio/detail/openstack-ardana-flavio/119/pipeline
- testbuild-gerrit: https://ci.suse.de/blue/organizations/jenkins/openstack-ardana-testbuild-gerrit-flavio/detail/openstack-ardana-testbuild-gerrit-flavio/18/pipeline/

File: scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py (hunk 1 above)
Comment (JanZerebecki) [reply scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py]: Please in a commit above the copyright claim add a link containing the commit hash to the upstream used as base. I didn't check, but maybe it is:
https://github.com/ansible/ansible/blob/a8d4bf86421d151d8df7132e8e87d04b6662f45a/lib/ansible/modules/cloud/openstack/os_stack.py

Do you mind if someone else submits the changes upstream?

Comment (JanZerebecki) [reply scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py]: To get the codeclimate check to not fail this, you also need to exclude this file.
Add a file named `.codeclimate.yml` to the top of the repo with the following content:
```
exclude_patterns:
- "scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py "
```

Comment (flaviodsr) [reply scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py]: > Please in a commit above the copyright claim add a link containing the commit hash to the upstream used as base. I didn't check, but maybe it is:
> [ansible/ansible:lib/ansible/modules/cloud/openstack/os_stack.py@`a8d4bf8`](https://github.com/ansible/ansible/blob/a8d4bf86421d151d8df7132e8e87d04b6662f45a/lib/ansible/modules/cloud/openstack/os_stack.py)
> 
> Do you mind if someone else submits the changes upstream?

no.


> To get the codeclimate check to not fail this, you also need to exclude this file.
> Add a file named `.codeclimate.yml` to the top of the repo with the following content:
> 
> ```
> exclude_patterns:
> - "scripts/jenkins/ardana/ansible/roles/heat_stack/library/ecp_os_stack.py "
> ```

I have added the file, thanks did not know about it :)

**Comment by flaviodsr** [comment]
> > break it at least in different commits, but I am afraid we cannot do more than this.
> 
> If you mean with "cannot", that it is technically impractical rather than that it is too time consuming, then that seems incorrect to me. Many of the dashes in your commit messages seem to indicate something that can be split into its own commit.
> 
> One example: the change related to ssh keys does not depend on anything else, right?

You are right, it is too time consuming.

File: scripts/jenkins/ardana/ansible/roles/heat_stack/tasks/delete.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        state: absent
+        api_timeout: "{{ api_timeout }}"
+      when: "'DELETE_COMPLETE' not in stack_status.stdout"
+  rescue:
+    - name: Force delete stack resources when delete failed
+      command: |
+         openstack --os-cloud {{ os_cloud }} stack resource list --filter \
+           type={{ item.type }} -f value -c physical_resource_id {{ heat_stack_name }} |
+           awk '{print "openstack --os-cloud '{{ os_cloud }}' {{ item.name }} \
+           delete --wait "$1 }' | sh -x || :
```
Comment (stefannica) [inline scripts/jenkins/ardana/ansible/roles/heat_stack/tasks/delete.yml]: '--wait' is only applicable for 'openstack server delete'

File: scripts/jenkins/ardana/ansible/roles/heat_stack/tasks/delete.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    - name: Retry delete stack
+      os_stack:
+        cloud: "{{ os_cloud }}"
+        name: "{{ heat_stack_name }}"
+        state: absent
+        api_timeout: "{{ api_timeout }}"
+      register: stack_status
+      retries: 3
+      until: "stack_status.rc == 0"
```
Comment (stefannica) [inline scripts/jenkins/ardana/ansible/roles/heat_stack/tasks/delete.yml]: One thing you **absolutely** want to avoid is delete a stack which is in a "DELETE_IN_PROGRESS" state. Doing this could result in un-delete-able stacks, such as the ones that you can see in a DELETE_FAILED status now. Which is why you need to check the stack's state every time before you try to delete it.

File: scripts/jenkins/ardana/ansible/roles/heat_stack/defaults/main.yml [hunk 4]
```diff
@@ -15,4 +15,8 @@
 #
 ---
 
-ses_rgw_port: "{{ hostvars['ardana-' + qe_env].ses_rgw_port }}"
+os_cloud: "engcloud-cloud-ci"
+heat_action: "create"
+api_timeout: "3600"
```
Comment (stefannica) [inline scripts/jenkins/ardana/ansible/roles/heat_stack/defaults/main.yml]: Could you set this back to 10 minutes, as it was the case with the old bash script this is replacing ?

File: jenkins/ci.suse.de/openstack-ardana-bootstrap-clm.yaml [hunk 5]
```diff
@@ -31,66 +31,38 @@
...
-          name: repositories
-          type: multi-select
-          value: SLES-Pool,SLES-Updates,SLES-Updates-test,Cloud-Pool,Cloud-Updates,Cloud-Updates-test
-          visible-items: 6
-          multi-select-delimiter: ','
-          default-value: SLES-Pool,SLES-Updates
-          description: >-
-            Set of zypper repositories (from provo-clouddata) to be configured during installation
-            in addition to that indicated by the cloudsource parameter.
+               GM<X>+up (official GM plus Cloud-Updates)
```
Comment (JanZerebecki) [inline jenkins/ci.suse.de/openstack-ardana-bootstrap-clm.yaml]: Please add an explanation what repos officialcloud9 uses.

File: jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            )
+          }
+          currentBuild.displayName = "#${BUILD_NUMBER} ${ardana_env}"
+        }
+      }
+    }
+
+    stage('Bootstrap CLM') {
+      steps {
+        ansible_playbook(
```
Comment (JanZerebecki) [inline jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile]: Why is this needed over "sh"?

Comment (JanZerebecki) [reply jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile]: I'd prefer moving this into a shell script in its own file.

File: scripts/jenkins/ardana/ansible/roles/heat_stack/tasks/delete.yml (hunk 3 above)
Comment (flaviodsr) [reply scripts/jenkins/ardana/ansible/roles/heat_stack/tasks/delete.yml]: As we discussed, the status of the stack should never be "DELETE_IN_PROGRESS" when it reaches this task.

File: scripts/jenkins/ardana/ansible/roles/heat_stack/defaults/main.yml (hunk 4 above)
Comment (flaviodsr) [reply scripts/jenkins/ardana/ansible/roles/heat_stack/defaults/main.yml]: done!

File: jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile (hunk 6 above)
Comment (JanZerebecki) [reply jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile]: How about doing that with a script in a file instead of inline where it can only be used from Jenkins?

File: jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+          if (ardana_env == '') {
+            error("Empty 'ardana_env' parameter value.")
+          }
+          if (reuse_workspace == '') {
+            sh('git clone $git_automation_repo --branch $git_automation_branch automation-git')
+            ansible_playbook(
+                playbook: "load-job-params.yml"
+            )
+          }
+          currentBuild.displayName = "#${BUILD_NUMBER} ${ardana_env}"
```
Comment (stefannica) [inline jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile]: Could you move this to be the first task in this stage immediately following the `ardana_env` empty check ? This will make sure the job name is set as early as possible.

File: jenkins/ci.suse.de/pipelines/openstack-ardana-pcloud.Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+          if (ardana_env == '') {
+            error("Empty 'ardana_env' parameter value.")
+          }
+          if (reuse_workspace == '') {
+            sh('git clone $git_automation_repo --branch $git_automation_branch automation-git')
+            ansible_playbook(
+                playbook: "load-job-params.yml"
+            )
+          }
+          currentBuild.displayName = "#${BUILD_NUMBER} ${ardana_env}"
```
Comment (stefannica) [inline jenkins/ci.suse.de/pipelines/openstack-ardana-pcloud.Jenkinsfile]: Same here

File: jenkins/ci.suse.de/pipelines/openstack-ardana-tempest.Jenkinsfile [hunk 9]
```diff
@@ -36,63 +37,52 @@ pipeline {
...
+            sh('git clone $git_automation_repo --branch $git_automation_branch automation-git')
+            ansible_playbook(
+              playbook: "load-job-params.yml"
+            )
+            ansible_playbook(
+              playbook: "setup-ssh-access.yml",
+              extra_vars: "-e @input.yml"
+            )
           }
+          currentBuild.displayName = "#${BUILD_NUMBER} ${ardana_env}"
```
Comment (stefannica) [inline jenkins/ci.suse.de/pipelines/openstack-ardana-tempest.Jenkinsfile]: and here

File: jenkins/ci.suse.de/pipelines/openstack-ardana-vcloud.Jenkinsfile [hunk 10]
```diff
@@ -38,147 +38,113 @@ pipeline {
...
+          if (ardana_env == '') {
             error("Empty 'ardana_env' parameter value.")
           }
+          if (reuse_workspace == '') {
+            sh('git clone $git_automation_repo --branch $git_automation_branch automation-git')
+            ansible_playbook(
+                playbook: "load-job-params.yml"
+            )
+          }
           currentBuild.displayName = "#${BUILD_NUMBER} ${ardana_env}"
```
Comment (stefannica) [inline jenkins/ci.suse.de/pipelines/openstack-ardana-vcloud.Jenkinsfile]: and here

File: jenkins/ci.suse.de/pipelines/openstack-ardana.Jenkinsfile [hunk 11]
```diff
@@ -16,328 +16,230 @@ pipeline {
...
         cleanWs()
-
         script {
           env.cloud_type = "virtual"
-          if ( ardana_env == '') {
+          if (ardana_env == '') {
             error("Empty 'ardana_env' parameter value.")
           }
-          currentBuild.displayName = "#${BUILD_NUMBER} ${ardana_env}"
           if ( ardana_env.startsWith("qe") || ardana_env.startsWith("qa") ) {
```
Comment (stefannica) [inline jenkins/ci.suse.de/pipelines/openstack-ardana.Jenkinsfile]: One last request, if you have the time for it: instead of using this technically incorrect check to determine if the the setup is hardware or virtual, could you check the `ardana_env` value against the ansible inventory using a shell command and save the result in the `cloud_type` environment variable ?

File: jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile (hunk 6 above)
Comment (flaviodsr) [reply jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile]: Some lines on the groovy function are not necessary when running outside jenkins. So, I am not sure if having a script for using outside jenkins would make sense. Personally, I prefer to just activate the venv and call `ansible-playbook`.

File: jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile (hunk 7 above)
Comment (flaviodsr) [reply jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile]: done.

File: jenkins/ci.suse.de/pipelines/openstack-ardana-pcloud.Jenkinsfile (hunk 8 above)
Comment (flaviodsr) [reply jenkins/ci.suse.de/pipelines/openstack-ardana-pcloud.Jenkinsfile]: done.

File: jenkins/ci.suse.de/pipelines/openstack-ardana-tempest.Jenkinsfile (hunk 9 above)
Comment (flaviodsr) [reply jenkins/ci.suse.de/pipelines/openstack-ardana-tempest.Jenkinsfile]: done.

File: jenkins/ci.suse.de/pipelines/openstack-ardana-vcloud.Jenkinsfile (hunk 10 above)
Comment (flaviodsr) [reply jenkins/ci.suse.de/pipelines/openstack-ardana-vcloud.Jenkinsfile]: done.

File: jenkins/ci.suse.de/pipelines/openstack-ardana.Jenkinsfile (hunk 11 above)
Comment (flaviodsr) [reply jenkins/ci.suse.de/pipelines/openstack-ardana.Jenkinsfile]: As we discussed, using pipe inside groovy `sh` is tricky and it will require a bit more time for testing. So, for now we are leaving it as it always has been.

File: jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile (hunk 6 above)
Comment (JanZerebecki) [reply jenkins/ci.suse.de/pipelines/openstack-ardana-bootstrap-clm.Jenkinsfile]: In chat Stefan suggested to source a `jenkins-helper.sh` for common functions to be called in Jenkinsfile sh steps. We agreed.
